python/lsst/afw/image/slicing.py

```diff
@@ -109,38 +130,18 @@ def clone(self):
     cls.clone = clone
 
     def __getitem__(self, imageSlice):
-        bbox, origin = _getBBoxFromSliceTuple(self, imageSlice)
-
-        _checkOrigin(origin)
-        return self.Factory(self, bbox, origin)
+        box, index, origin = interpretSliceArgs(imageSlice, self.getBBox)
+        if box is not None:
+            return self.subset(box, origin=origin)
+        return self._get(index, origin=origin)
     cls.__getitem__ = __getitem__
 
     def __setitem__(self, imageSlice, rhs):
-        bbox, origin = _getBBoxFromSliceTuple(self, imageSlice)
-
-        _checkOrigin(origin)
-        if self.assign(rhs, bbox, origin) is NotImplemented:
-            lhs = self.Factory(self, bbox, origin=origin)
-            lhs.set(rhs)
+        box, index, origin = interpretSliceArgs(imageSlice, self.getBBox)
+        if box is not None:
+            if self.assign(rhs, box, origin) is NotImplemented:
+                lhs = self.subset(box, origin=origin)
+                lhs.set(rhs)
```

Comment:
As we discussed in regard to `set` in multiband images, `set` can probably be removed and this line would just be `lhs[:] = rhs`.

Reply:
That would recurse infinitely, right?  I think that would invoke `__setitem__` on `lhs`, which would be this same method on a different instance of the same class.

I think the question is whether this `set` should also be renamed to `_set`, and I'll admit that the only reason I don't want to do that now is that it'll be another breaking change that will require fixing a lot more downstream code.

Reply:
Sorry, you're right. I think I meant `lhs.array[:] = rhs` (or `lhs.array[:] = rhs.array if rhs is an `Image`). I was just thinking of our discussion in DM-13911 about how having both `set` and `__setiitem__` added unnecessary code to maintain and feel like if we remove it from the multiband classes then we should also remove it from the single band python API.

Reply:
Unfortunately `lhs.array[:]` doesn't work for `MaskedImage` or `Exposure`, so I think we need some other (at least internal) method for those classes.  It'd be ideal if `assign` just took care of this case (I'm actually not sure it doesn't - this is probably the only piece of code in `__setitem__` that I left as it was, though you can't tell that from the diff).

I suppose that means we might want something for the multiband classes, then, too, but note that this `set` modifies all pixels in the (generalized) image, not individual ones.  I also tend to think that the advantages of keeping interfaces lean are often greater than the advantages of making them the same, so maybe the multiband classes don't need them if we don't expect to need to modify those objects that way.  But I suppose that's a conversation for the other ticket.

For this one, mind if we agree that making this `set` private or folding it into `assign` is a good idea, but do not try to fix it on this ticket?

Reply:
Sure. I don't really have strong feelings about this if you don't mind the API's being slightly different from the multiband classes. So you can just leave it as is.